src/move.c

```diff
@@ -3161,35 +3203,46 @@ pagescroll(int dir, long count, int half)
 	if (count)
 	    curwin->w_p_scr = MIN(curwin->w_height, count);
 	count = MIN(curwin->w_height, curwin->w_p_scr);
+	int curscount = count;
 
-	// Don't scroll if we already know that it will reveal end of buffer lines.
-	if (dir == BACKWARD
-	    || (curwin->w_botline - 1 < buflen)
-	    || (curwin->w_p_sms && curwin->w_botline - 1 == buflen
-		&& curwin->w_skipcol < linetabsize(curwin, buflen)))
+	// Adjust count so as to not reveal end of buffer lines.
+	if (dir == FORWARD)
 	{
-	    nochange = scroll_with_sms(dir, &count);
-	    validate_botline();
-	    // Hide any potentially revealed end of buffer lines.
-	    if (!nochange && curwin->w_botline - 1 == buflen)
+	    int n = plines_win(curwin, curwin->w_topline, FALSE);
+		n -= adjust_plines_for_skipcol(curwin);
+	    if (n - count < curwin->w_height && curwin->w_topline < buflen)
+		n += plines_m_win(curwin, curwin->w_topline + 1, buflen, TRUE);
```

Comment:
This does not include filler lines below the last line in the buffer. Thus Ctrl-D near the end of a file does not reveal any potential filler lines (nor did it before the inital halfpage smoothscroll patch). I think it would make sense if pressing Ctrl-D near the end of a file would reveal filler lines. A low-level plines fix for that could be:
```diff
@@ -347,8 +347,11 @@ plines_win(
 #if defined(FEAT_DIFF) || defined(PROTO)
     // Check for filler lines above this buffer line.  When folded the result
     // is one line anyway.
-    return plines_win_nofill(wp, lnum, limit_winheight)
+    int n = plines_win_nofill(wp, lnum, limit_winheight)
                                                   + diff_check_fill(wp, lnum);
+    if (lnum == wp->w_buffer->b_ml.ml_line_count)
+       n += diff_check_fill(wp, lnum + 1);
+    return n;
 }

 /*
```

I think it makes sense because plines() can never be called with lnum == b_ml.ml_line_count + 1, which is where filler lines below the last line are attached to.

Otherwise I we could handle it in pagescroll() directly.

Reply:
I don't think the change should be in `plines_win()` as that's inconsistent and may lead to more mistakes.

Reply:
Fair enough. Do you think adding it to plines_m_win() is fine? Otherwise I may need to duplicate it's logic in pagescroll(). Or change plines_m_win "last" argument to a pointer whose final value can be checked at the callsite.

Reply:
I assume yes and assume the behavior is wanted and update the PR, let me know if it is not.

Reply:
I think adding it to `plines_m_win()` makes sense, considering that it also uses `w_topfill` for `w_topline`.